pulp_2to3_migration/app/pre_migration.py

```diff
@@ -355,8 +355,45 @@ def pre_migrate_all_without_content(plan):
             # filter by repo type
             repos_to_check = plan.type_to_repo_ids[plugin_plan.type]
 
-            mongo_repo_q = mongo_Q(repo_id__in=repos_to_check)
-            mongo_repo_qs = Repository.objects(mongo_repo_q)
+            # figure out which repositories/importers/distributors were updated since the last run
+            repo_premigrated_last = Pulp2Repository.objects.aggregate(
+                Max('pulp_last_updated')
+            )['pulp_last_updated__max'] or 0
+            imp_premigrated_last = Pulp2Importer.objects.aggregate(
+                Max('pulp_last_updated')
+            )['pulp_last_updated__max'] or 0
+            dist_premigrated_last = Pulp2Distributor.objects.aggregate(
+                Max('pulp_last_updated')
+            )['pulp_last_updated__max'] or 0
+
+            is_content_added_q = mongo_Q(last_unit_added__gte=repo_premigrated_last)
+            is_content_removed_q = mongo_Q(last_unit_removed__gte=repo_premigrated_last)
+            is_new_enough_repo_q = is_content_added_q | is_content_removed_q
+            is_empty_repo_q = (
+                mongo_Q(last_unit_added__exists=False) & mongo_Q(last_unit_removed__exists=False)
+            )
+            is_new_enough_imp_q = mongo_Q(last_updated__gte=imp_premigrated_last)
+            is_new_enough_dist_q = mongo_Q(last_updated__gte=dist_premigrated_last)
+            repo_id_q = mongo_Q(repo_id__in=repos_to_check)
+
+            updated_importers = Importer.objects(
+                repo_id_q & is_new_enough_imp_q
+            ).only('repo_id')
+            updated_imp_repos = set(imp.repo_id for imp in updated_importers)
```

Comment:
what is the need of making a set out of it?

Reply:
Probably no need. We need to have a sequence containing repo_id and a set is a safe way to ensure they are unique.
I agree that importers are not reused in pulp2, so should be unique regardless. I can switch it to tuple or list if it's preferable.

Reply:
we can just .distinct('repo_id') this will give you a list with unique rpeo_ids

Reply:
i am still suggesting to use distinct in this case and not iterating through the qs
```
d=Distributor.objects.distinct('repo_id')
i=Importer.objects.distinct('repo_id')
result = set().union(d,i)
```

Reply:
@ipanova , I'm afraid of mongo's distinct, it can cause OOM easily, not sure how much data is needed for it. I experienced it in the past, with applicability specifically. I'd do as little heavy operations as possible in mongo anyway since it's running in production and loaded with other stuff.

Reply:
@goosemania that's a fair point.